Firmware/fibre/python/fibre/discovery.py

```diff
@@ -64,25 +67,71 @@ def did_discover_channel(channel):
         """
         try:
             logger.debug("Connecting to device on " + channel._name)
+
+            temp_dir = tempfile.gettempdir()
+
+            cache_miss = True
+            json_crc16 = 0
+
+            # Fetching the json crc to check cache
             try:
-                json_bytes = channel.remote_endpoint_read_buffer(0)
-            except (TimeoutError, ChannelBrokenException):
-                logger.debug("no response - probably incompatible")
-                return
+                json_crc16 = channel.remote_endpoint_operation(0, struct.pack("<I", 0xffffffff), True, 2)
+                json_crc16 = struct.unpack("<H", json_crc16)[0]
+                logger.debug("Device JSON checksum: 0x{:02X} 0x{:02X}".format(json_crc16 & 0xff, (json_crc16 >> 8) & 0xff))
+            except Exception as error:
+                logger.debug("Error fetching JSON CRC, falling back to downloading full JSON")
+                #logger.debug(traceback.format_exc())
+
+            cache_path =  temp_dir + '/fibre_schema_cache_' + str(json_crc16)
+
+            try:
+                json_cache = open(cache_path, 'r+')
+                logger.debug("Found cache file with crc")
+                cache_miss = False
+            except:
+                logger.debug("Cache miss, there is no cache file")
+                cache_miss = True
```

Comment:
Can you merge this try block with the one above, and hence get rid of json_crc16 = 0 ?
If you want to have a default for old firmware, you need to put it in the except block, or it may be ill-formated.